### PR TITLE
Add expand_coverage_report option to for #270 

### LINKED ISSUE #270: False negatives for function calls splitted along multiple lines
Hi, thanks for this library. It is very helpful in our project!

I think I found an issue, false negative, that affects overall quality of diff-cover check in a specific use case.

I use diff-cover 6.4.4 with coverage.py in version 6.2 through pytest-cov in version 3.0.0 with Cobertura XML report.

* Recently I updated function call in my production code (the signature of the function I use changed, it has a new argument, and I've added new keyword argument to the call)
* This production call has no test coverage.
* Diff-cover doesn't fail for the line with a new argument in the function call.

Expected:
* Diff-cover fails for the line with a new argument in function call because of lack of the coverage.

Details: I think that probably diff-cover ignores lines without coverage report, and probably (?) it should for every changed line in diff search for first line above with reported coverage. On the image below red line point to lines not covered by tests (GitLab supports Cobertura reports).

Zrzut ekranu 2022-01-17 o 12 33 36
Zrzut ekranu 2022-01-17 o 12 33 49

```
$ diff-cover coverage.xml --fail-under 100 --html-report diff-cover.html --compare-branch origin/develop
-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
…/file_on_screenshot.py (100%)
…/obfuscated.py (100%)
…/obfuscated2.py (100%)
-------------
Total:   13 lines
Missing: 0 lines
Coverage: 100%
-------------
```

### PR DESCRIPTION
**Pull request by MathieuLamiot**
# Description

## Issue
This PR provides a work-around that fixes #270 
The issue reported there is that diff data contains all modified lines, while most coverage tools report by statement and not lines.
As an example, let's take this code change:

```
429        if worker is None:
430            metrics.UNEXPECTED_RETURN_VALUE_COUNTER.labels(
- 431                code=self._UNKNOWN_WORKER_RESPONSE["returnvalue"]["code"],
+ 431                code=self._UNKNOWN_WORKER_RESPONSE["code"],
432                method=request.method,
433                endpoint=request.path,
434            ).inc()
435            return Response(self._UNKNOWN_WORKER_RESPONSE)
```

Diff file will report line 431 as being changed.
However, most coverage tools will report lines 429, 430 and 435 being hit or not by test, but nothing for line 431.

As a result, the Violation reporters do not report line 431 has a violation, even if it does not run during tests.
But diff reports don't report line 430 has being changed. So in the end, this changes goes silent in the diff-cover analysis.

## Fix implemented

Following this suggestion: https://github.com/Bachmann1234/diff_cover/issues/270#issuecomment-1017563602, it seems most coverage reports use the line opening a statement as the reference. Also, untested statements are reported in the coverage files, with 0 hits (they are not just missing).

The implemented solution is to:
- add an option --expand-coverage-report (not on by default because this workaround might not work in all situations, to be further tested with different coverage reporters and languages.)
- When the option is set, all line numbers not listed in the XML reports are extrapolated from the previous line reported. 

In the example above, line 431 is not in coverage.xml. But line 430 is, so we add line 431 in the data extracted from the XML report, with the same number of hits as 430.

## Risks & limitations

- The option is only implemented for XML reports.
- This fix works as expected under the assumptions that:
    - The coverage tool reports untested statements too, with hits set to 0.
    - The coverage tool reports statements based on the opening line.
- Blank lines/empty lines might be reported?
- There are no unit/integration tests

## Tests performed

I tested this PR on the example presented above (so in Python), and with other changes covered in the same file ; with:
- `pytest --cov=. --cov-report=xml `
- `diff-cover coverage.xml --compare-branch=origin/develop --markdown-report diff-cover-report.md --exclude 'test*.py' --fail-under=50 --expand_coverage_report`

### When line 430 is not tested

- coverage.xml contains

```
						<line number="424" hits="0"/>
						<line number="426" hits="1"/>
						<line number="427" hits="1"/>
						<line number="428" hits="1"/>
						<line number="429" hits="1"/>
						<line number="430" hits="0"/>
						<line number="435" hits="0"/>
						<line number="436" hits="1"/>
						<line number="438" hits="1"/>
						<line number="439" hits="1"/>
```

and diff-coverage returns:

-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
saas/views.py (95.5%): Missing lines 431
-------------
Total:   22 lines
Missing: 1 line
Coverage: 95%
-------------
The markdown:

> # Diff Coverage
> ## Diff: origin/develop...HEAD, staged and unstaged changes
> 
> - saas/views&#46;py (95.5%): Missing lines 431
> 
> ## Summary
> 
> - **Total**: 22 lines
> - **Missing**: 1 line
> - **Coverage**: 95%
> 
> 
> 
> ## saas/views&#46;py
> 
> Lines 427-435
> 
> ```python
>   427         params["id"] = job_id
>   428         worker = settings.WORKERS.get(worker_name.lower())
>   429         if worker is None:
>   430             metrics.UNEXPECTED_RETURN_VALUE_COUNTER.labels(
> ! 431                 code=self._UNKNOWN_WORKER_RESPONSE["code"],
>   432                 method=request.method,
>   433                 endpoint=request.path,
>   434             ).inc()
>   435             return Response(self._UNKNOWN_WORKER_RESPONSE)
> ```
> 
> 
> ---

### When line 431 is tested

- coverage.xml contains

```
						<line number="424" hits="0"/>
						<line number="426" hits="1"/>
						<line number="427" hits="1"/>
						<line number="428" hits="1"/>
						<line number="429" hits="1"/>
						<line number="430" hits="1"/>
						<line number="435" hits="0"/>
						<line number="436" hits="1"/>
						<line number="438" hits="1"/>
						<line number="439" hits="1"/>
```

and diff-coverage returns:

-------------
Diff Coverage
Diff: origin/develop...HEAD, staged and unstaged changes
-------------
saas/views.py (100%)
-------------
Total:   22 lines
Missing: 0 lines
Coverage: 100%
-------------

The markdown:

> # Diff Coverage
> ## Diff: origin/develop...HEAD, staged and unstaged changes
> 
> - saas/views&#46;py (100%)
> 
> ## Summary
> 
> - **Total**: 22 lines
> - **Missing**: 0 lines
> - **Coverage**: 100%
>

### CODE REVIEW

**Comment by Bachmann1234** [comment]
I'm gonna need a bit of time before I get to this. But I see it. Just a busy time this week.

**Comment by Bachmann1234** [comment]
This is pretty interesting. I feel a bit iffy about trying to interpret data not in the report. But this seems sensible. 

The flag should be documented in the readme
and ill need a few tests before I can merge this

**Comment by MathieuLamiot** [comment]
Thank you for the feedback. I won't have time right now for the test but I'll try to do something about this in the upcoming weeks for sure!

File: README.rst [hunk 1]
```diff
@@ -274,6 +274,18 @@ It can be enabled by using the ``-q``/``--quiet`` flag:
 
 If enabled, the tool will only print errors and failures but no information or warning messages.
 
+Compatibility with multi-line statements 
+----------
```
Comment (Bachmann1234) [inline README.rst:278]:
```suggestion
Compatibility with multi-line statements 
----------------------------------------
```

File: README.rst [hunk 2]
```diff
@@ -274,6 +274,18 @@ It can be enabled by using the ``-q``/``--quiet`` flag:
 
 If enabled, the tool will only print errors and failures but no information or warning messages.
 
+Compatibility with multi-line statements
```
Comment (Bachmann1234) [inline README.rst:277]:
```suggestion
Compatibility with multi-line statements
```

**Comment by Bachmann1234** [comment]
Annoyingly my readme formatting check had some nitpicks, I made the suggestions that I __think__ will fix these. 

Pr looks good, once we get this merged ill try to get it out there asap.

**Comment by MathieuLamiot** [comment]
Good catch, thanks. It should be good for the 2 feedbacks you shared now 👍

**Comment by Bachmann1234** [comment]
Alright. Let's merge this in. I'll get it released in the next day or so

**Comment by Bachmann1234** [comment]
Thank you! Try out https://pypi.org/project/diff-cover/9.2.0/

**Pull request merged**